Move handshake queue and workers into run()

zslehoczky opened the pull request:
Plus: in tests.rs, add helper function for wireguard creation instead of duplicated code

zslehoczky committed:
Move handshake workers into run

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Refactor device creation in tests

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

rot256 commented on wireguard-rs/src/run/run.rs:
```diff
@@ -50,11 +51,19 @@ fn run(config: Config) -> Result<(), ErrorReason> {
 
     thread::scope(|thread_scope| {
+        // start handshake workers
+        for _ in 0..num_cpus::get() {
```
We can use:

https://doc.rust-lang.org/std/thread/fn.available_parallelism.html

And get rid of `num_cpus` now.

zslehoczky replied:
Moved it to dev-dependencies, because we use it in bench.rs to get the number of physical cores, which is not the same as available_parallelism I think. But switched to the new function at other places

rot256 reviewed (dismissed):
This is cleaner!

zslehoczky committed:
Merge remote-tracking branch 'origin/develop' into queue

zslehoczky committed:
Move num_cpu dependency to dev

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Fix merge error

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Rename generic channel variable names

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Small updates with channel creation

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Pass number of cpu-s in ctor

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Fix: close handshake queue before joining threads

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Create handshake worker spawn fn

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky commented:
I fixed the review suggestion, tidied it up a little and also fixed a logic error: handshake sender was never closed so the loop on the receiver never exited

zslehoczky committed:
Merge remote-tracking branch 'origin/develop' into queue

zslehoczky committed:
Apply suggestion

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky committed:
Apply fmt

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>

zslehoczky commented:
@rot256 Also updated this one with the latest, so this is also ready for review

rot256 commented on Cargo.toml:
```diff
@@ -62,4 +61,5 @@ proptest = "^0.10"
 rand_chacha = "0.3"
 bencher = "0.1.5"
 chacha20 = "0.9"
+num_cpus = "^1.10"
```
We are still using this elsewhere?

zslehoczky replied:
We use it in bench.rs:
`num_cpus::get_physical()`

get_physical is not the same as get, so I left it as is

zslehoczky replied:
But looking at it, I think it would make sense to just use the same number in bench.rs as everywhere else

zslehoczky replied:
Changed it

rot256 commented on wireguard-rs/src/wireguard/tests.rs:
```diff
@@ -75,6 +77,31 @@ fn init() {
+    let (sender, receiver) = bounded(n_cpus);
+
+    let (fake, tun_reader, tun_writer, _) = dummy::TunTest::create(true);
+    let wireguard_device: WireGuard<dummy::TunTest, dummy::PairBind> =
+        WireGuard::new(tun_writer, sender, n_cpus);
```
The workers taking elements from `sender` are still spawned inside `WireGuard`?

zslehoczky replied:
No, sender is the input side of the channel and it's used to send the jobs to the worker. The worker thread is in runner, and it has the receiver side

rot256 reviewed (dismissed):
Looks good

zslehoczky committed:
Remove num_cpus completely

Signed-off-by: Zsombor Lehoczky <zslehoczky@users.noreply.github.com>
